KEYCLOAK-14113 Support for exchanging to SAML 2.0 token

wadahiro opened the pull request:
Fix for https://issues.redhat.com/browse/KEYCLOAK-14113.

wadahiro committed:
KEYCLOAK-14113 Support for exchanging to SAML 2.0 token

pedroigor reviewed (changes requested):
@wadahiro Looks very good to me.

Could you just revert the changes where you are basically formatting the code? It is a bit of nitpicking, but it makes the review process easier. In the future we should have a code style to avoid such confusion.

wadahiro commented:
@pedroigor 

> Could you just revert the changes where you are basically formatting the code? It is a bit of nitpicking, but it makes the review process easier.

Ok. I reverted them. Also, I rebased onto the latest master branch.

pedroigor commented on saml-core/src/main/java/org/keycloak/saml/common/util/DocumentUtil.java:
```diff
@@ -299,6 +299,33 @@ public static InputStream getSourceAsStream(Source source) throws ConfigurationE
+     * @throws TransformerFactoryConfigurationError
+     * @throws TransformerException
+     */
+    public static String getNodeAsString(Node node) throws ConfigurationException, ProcessingException {
```
I just noticed that this method org.keycloak.saml.common.util.DocumentUtil#getDocumentAsString does the same thing.

Does it work for you?

wadahiro replied:
@pedroigor 
We can't call `DocumentUtil#getDocumentAsString` directly because of the type miss match of the argument. How about refactoring the `getDocumentAsString` method and adding `getNodeAsString` method as follows?

```
    public static String getDocumentAsString(Document signedDoc) throws ProcessingException, ConfigurationException {
        return getNodeAsString(signedDoc);
    }

    public static String getNodeAsString(Node node) throws ConfigurationException, ProcessingException {
        Source source = new DOMSource(node);
        StringWriter sw = new StringWriter();

        Result streamResult = new StreamResult(sw);
        // Write the DOM document to the stream
        Transformer xformer = TransformerUtil.getTransformer();
        try {
            xformer.transform(source, streamResult);
        } catch (TransformerException e) {
            throw logger.processingError(e);
        }

        return sw.toString();
    }
```

pedroigor replied:
It works. I think that is what I was trying to say but I forgot you are actually using a Node and not a Document.

pedroigor commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java:
```diff
@@ -945,6 +976,56 @@ protected Response exchangeClientToClient(UserModel targetUser, UserSessionModel
+
+        int assertionLifespan = samlClient.getAssertionLifespan();
+
+        AccessTokenResponse res = new AccessTokenResponse();
```
Could you include `issued_token_type` and set it to `urn:ietf:params:oauth:token-type:saml2`. See https://tools.ietf.org/id/draft-ietf-oauth-token-exchange-19.html#rfc.section.3.

I guess that would mean creating a `TokenExchangeResponse` extending `AccessTokenResponse` ? Reason being that we use that new class only for token exchange and don't touch the oauth2 part. Also easier to support additional parameters in the future that could be specific of exchange.

wadahiro replied:
@pedroigor 

> Could you include issued_token_type and set it to `urn:ietf:params:oauth:token-type:saml2`

It's already included in the following code through `AccessTokenResponse#setOtherClaims` method:
https://github.com/keycloak/keycloak/pull/7064/files#diff-ec5f9626f53934cdc3ad494e198f1902R1021

But there is no assertion in the test code. I'll add the assertion.

> I guess that would mean creating a `TokenExchangeResponse` extending `AccessTokenResponse`

- Create `TokenExchangeResponse` extending `AccessTokenResponse` as new class
- Add `issuedTokenType` field and set/get method into the new class
- Fix the existing token exchange code for OIDC client and external clients (implementation of `ExchangeTokenToIdentityProviderToken`)to use the new response class

Is the above fix OK?

Also, I noticed the current token exchange code doesn't follow the spec.

- Exchange for OIDC client doesn't include `issued_token_type` claim.
- Refresh token for OIDC client isn't stored in `access_token` claim when `requested_token_type` is refresh token type. 
https://github.com/keycloak/keycloak/blob/6dde131609551621eba102d71c5897a350d7ce64/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java#L930-L933
- ID token is issued by the `scope` parameter. Should we issue the id token when `requested_token_type` is id token type?
https://github.com/keycloak/keycloak/blob/6dde131609551621eba102d71c5897a350d7ce64/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java#L935-L938
- ID token for external client isn't stored in `access_token` claim when `requested_token_type` is id token type. 
https://github.com/keycloak/keycloak/blob/7efaf9869acbb328f9dd8b159460cec7de6bc010/services/src/main/java/org/keycloak/broker/oidc/AbstractOAuth2IdentityProvider.java#L225-L230

Should I fix those issues in this pull request too?

pedroigor replied:
Yeah, we are missing some bits of the specs. I really missed the part you are setting the `issued_token_type`.

Don't worry about fixing them now. Let's do it in a separate and specific PR.

wadahiro replied:
All right, I'll fix `DocumentUtil` and test case.

pedroigor commented:
@wadahiro Sorry ... I have these last comments, which I did not catch before. Does it make sense for you?

I'm done with the review :)

pedroigor commented:
@wadahiro I think the last bit is the documentutil stuff. Once you do that, we should just merge.

wadahiro commented:
@pedroigor I updated my commit. Also, I rebased onto the latest master branch again. Could you check it again?

pedroigor commented:
@wadahiro awesome, thanks.
